File: ios/KaasSdk.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(KaasSdk, NSObject)

RCT_EXTERN_METHOD(sendCommand:(NSString *)command
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(createSessionRequestToken:(BOOL *)enableUserAuthenticatedVirtualKeys
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(openSession:(NSDictionary *)sessionResponse
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(closeSession:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isSessionOpened:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(createVirtualKey:(NSDictionary *)sessionResponse
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(updateVirtualKey:(NSDictionary *)updateVirtualKeyRequest
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(revokeVirtualKey:(NSDictionary *)revokeVirtualKeyRequest
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getVirtualKeys:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(selectVirtualKey:(NSDictionary *)selectVirtualKeyRequest
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getSelectedVirtualKey:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(scan:(NSNumber *)timeoutInSeconds
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(connect:(NSNumber *)timeout
                  disableTimeSync:(BOOL *)disableTimeSync
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(bluetoothStateChanged:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isConnected:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(disconnect:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getVehicleData:(NSString *)dataItem
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: ios/KaasSdk.swift
@objc(KaasSdk)
class KaasSdk: NSObject {
    
    @objc(sendCommand:withResolver:withRejecter:)
    func sendCommand(command: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        resolve("SUCCESSFUL")
    }
    
    @objc(createSessionRequestToken:withResolver:withRejecter:)
    func createSessionRequestToken(enableUserAuthenticatedVirtualKeys: Bool, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        let timestamp = NSDate().timeIntervalSince1970
        let result = [
          "sessionRequestToken": timestamp
        ]
        resolve(result)
    }
    
    @objc(openSession:withResolver:withRejecter:)
    func openSession(sessionResponse: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("openSession", sessionResponse["sessionResponseToken"] ?? "-")
        resolve("success")
    }
    
    @objc
    func closeSession(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("closeSession")
        resolve(true)
    }
    
    @objc
    func isSessionOpened(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("isSessionOpened")
        resolve(true)
    }
    
    @objc(createVirtualKey:withResolver:withRejecter:)
    func createVirtualKey(createVirtualKeyRequest: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("createVirtualKey", createVirtualKeyRequest["sharedDeviceId"] ?? "-")
        let timestamp = NSDate().timeIntervalSince1970
        let result = [
          "sessionRequestToken": timestamp,
          "vehicleId": "vehicleId",
          "bluetoothName": "bluetoothName",
          "virtualKeyVersion": 1.22,
          "serialNumber": 122.3,
          "validityStartDate": NSDate().timeIntervalSince1970,
          "validityEndDate": NSDate().timeIntervalSince1970,
          "enableUserAuthenticated": true,
          "actionsAllowedMask": "actionsAllowedMask",
          "numberOfActionsAllowed": 5
        ] as [String : Any]
        resolve(result)
    }
    
    @objc(updateVirtualKey:withResolver:withRejecter:)
    func updateVirtualKey(updateVirtualKeyRequest: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("updateVirtualKey", updateVirtualKeyRequest["id"] ?? "-")
        let timestamp = NSDate().timeIntervalSince1970
        let result = [
          "id": timestamp,
          "vehicleId": "vehicleId",
          "bluetoothName": "bluetoothName",
          "virtualKeyVersion": 1.22,
          "serialNumber": 122.3,
          "validityStartDate": NSDate().timeIntervalSince1970,
          "validityEndDate": NSDate().timeIntervalSince1970,
          "enableUserAuthenticated": true,
          "actionsAllowedMask": "actionsAllowedMask",
          "numberOfActionsAllowed": 5
        ] as [String : Any]
        resolve(result)
    }
    
    @objc(revokeVirtualKey:withResolver:withRejecter:)
    func revokeVirtualKey(revokeVirtualKeyRequest: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("revokeVirtualKey", revokeVirtualKeyRequest["id"] ?? "-")
        resolve("OK")
    }
    
    @objc
    func getVirtualKeys(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("getVirtualKeys")
        let timestamp = NSDate().timeIntervalSince1970
        let item = [
          "id": timestamp,
          "vehicleId": "vehicleId",
          "bluetoothName": "bluetoothName",
          "virtualKeyVersion": 1.22,
          "serialNumber": 122.3,
          "validityStartDate": NSDate().timeIntervalSince1970,
          "validityEndDate": NSDate().timeIntervalSince1970,
          "enableUserAuthenticated": true,
          "actionsAllowedMask": "actionsAllowedMask",
          "numberOfActionsAllowed": 5
        ] as [String : Any]
        
        let result: NSMutableArray = []
        result.add(item)
        resolve(result)
    }
    
    @objc(selectVirtualKey:withResolver:withRejecter:)
    func selectVirtualKey(selectVirtualKeyRequest: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("selectVirtualKey", selectVirtualKeyRequest["id"] ?? "-")
        let timestamp = NSDate().timeIntervalSince1970
        let result = [
          "id": timestamp,
          "vehicleId": "vehicleId",
          "bluetoothName": "bluetoothName",
          "virtualKeyVersion": 1.22,
          "serialNumber": 122.3,
          "validityStartDate": NSDate().timeIntervalSince1970,
          "validityEndDate": NSDate().timeIntervalSince1970,
          "enableUserAuthenticated": true,
          "actionsAllowedMask": "actionsAllowedMask",
          "numberOfActionsAllowed": 5
        ] as [String : Any]
        resolve(result)
    }
    
    @objc
    func getSelectedVirtualKey(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("getSelectedVirtualKey")
        let timestamp = NSDate().timeIntervalSince1970
        let result = [
          "id": timestamp,
          "vehicleId": "vehicleId",
          "bluetoothName": "bluetoothName",
          "virtualKeyVersion": 1.22,
          "serialNumber": 122.3,
          "validityStartDate": NSDate().timeIntervalSince1970,
          "validityEndDate": NSDate().timeIntervalSince1970,
          "enableUserAuthenticated": true,
          "actionsAllowedMask": "actionsAllowedMask",
          "numberOfActionsAllowed": 5
        ] as [String : Any]
        
        resolve(result)
    }
    
    @objc(scan:withResolver:withRejecter:)
    func scan(timeoutInSeconds: Int, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("scan", timeoutInSeconds)
        resolve("VISIBLE")
    }
    
    @objc(connect:disableTimeSync:withResolver:withRejecter:)
    func connect(timeout: Int, disableTimeSync: Bool, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("connect", timeout, disableTimeSync)
        resolve("SUCCESSFUL")
    }
    
    @objc
    func bluetoothStateChanged(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("bluetoothStateChanged")
        resolve("READY")
    }
    
    @objc
    func isConnected(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("isConnected")
        resolve(true)
    }
    
    @objc
    func disconnect(_ resolve: RCTPromiseResolveBlock, withRejecter reject: RCTPromiseRejectBlock) -> Void {
        print("disconnect")
        resolve(true)
    }
    
    @objc(getVehicleData:withResolver:withRejecter:)
    func getVehicleData(dataItem: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("getVehicleData", dataItem)
        let timestamp = NSDate().timeIntervalSince1970
        let result = [
          "odometer": [
            "value": 2.3,
            "time": timestamp,
            "unit": "br"
          ]
        ] as [String : Any]
        
        resolve(result)
    }
}
